homeassistant/components/generic_thermostat/climate.py

```diff
@@ -44,428 +46,813 @@
+                "Cool mode enabled; _target_temp_cool: %s, entity_id: %s",
+                self._target_temp_cool,
+                self._ac_entity_id,
+            )
+        else:
+            self._cool_conf = None
+            self._target_temp_cool = None
+            self._ac_entity_id = None
+
+        self._current_temperature = None
         self._temp_lock = asyncio.Lock()
-        self._min_temp = min_temp
-        self._max_temp = max_temp
-        self._target_temp = target_temp
-        self._unit = unit
-        self._support_flags = SUPPORT_FLAGS
-        if away_temp:
-            self._support_flags = SUPPORT_FLAGS | SUPPORT_PRESET_MODE
-        self._away_temp = away_temp
-        self._is_away = False
 
     async def async_added_to_hass(self):
-        """Run when entity about to be added."""
+        """Run when entity about to be added.
+
+        Attach the listeners.
+        """
         await super().async_added_to_hass()
 
-        # Add listener
-        async_track_state_change(
-            self.hass, self.sensor_entity_id, self._async_sensor_changed
-        )
+        # Add listeners to track changes from the sensor and the heater's switch
         async_track_state_change(
-            self.hass, self.heater_entity_id, self._async_switch_changed
+            self.hass, self._sensor_entity_id, self._async_sensor_temperature_changed
         )
+        if self._is_heat_enabled:
+            async_track_state_change(
+                self.hass, self._heater_entity_id, self._async_switch_device_changed
+            )
+        if self._is_cool_enabled:
+            async_track_state_change(
+                self.hass, self._ac_entity_id, self._async_switch_device_changed
+            )
 
-        if self._keep_alive:
+        if self._sensor_stale_duration:
             async_track_time_interval(
-                self.hass, self._async_control_heating, self._keep_alive
+                self.hass,
+                self._async_check_sensor_not_responding,
+                self._sensor_stale_duration,
             )
 
+        if self._keep_alive:
+            async_track_time_interval(self.hass, self._async_operate, self._keep_alive)
+
         @callback
         def _async_startup(event):
             """Init on startup."""
-            sensor_state = self.hass.states.get(self.sensor_entity_id)
-            if sensor_state and sensor_state.state not in (
-                STATE_UNAVAILABLE,
-                STATE_UNKNOWN,
-            ):
-                self._async_update_temp(sensor_state)
+            sensor_state = self.hass.states.get(self._sensor_entity_id)
+            if sensor_state and sensor_state.state != STATE_UNKNOWN:
+                self._async_update_current_temp(sensor_state)
+                self.async_write_ha_state()
 
         self.hass.bus.async_listen_once(EVENT_HOMEASSISTANT_START, _async_startup)
 
-        # Check If we have an old state
+        _async_startup(None)  # init the sensor
```

Comment:
Shouldn't we call it after we restore the state? The function calls `self._async_update_current_temp` which might need the correct state.